### PR TITLE
fix: take snapshot if nothing was exported since last snapshot

### LINKED ISSUE #9761: Failed to take snapshot in leader because index entry is not found
**Describe the bug**

```
2022-07-12 01:28:37.329 CEST
Failed to take a snapshot for Broker-1-StreamProcessor-2
"io.camunda.zeebe.broker.system.partitions.NoEntryAtSnapshotPosition: Failed to take snapshot. Expected to find an indexed entry for determined snapshot position 1494387075 (processedPosition = 1494736551, exportedPosition=1494387075), but found no matching indexed entry which contains this position.
```

This occurred in the leader.

Related to #7911 . But in that issue, it was followers who could not take snapshot and that is a valid scenario. However, leaders should always be able to take snapshot.

The impact of this bug is low (See following section). There is no data inconsistency or loss. The snapshot is taken later when new events are exported. 

**To Reproduce**

Following events happened in the broker leading to the above error.

1. Broker receives a new snapshot and replace existing log.
```
 2022-07-12 01:26:19.338 CEST
RaftServer{raft-partition-partition-2}{role=FOLLOWER} - Committed snapshot FileBasedSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/2/snapshots/494998838-19-1494391974-1494387075, checksumFile=/usr/local/zeebe/data/raft-partition/partitions/2/snapshots/494998838-19-1494391974-1494387075.checksum, checksum=1722253646, metadata=FileBasedSnapshotMetadata{index=494998838, term=19, processedPosition=1494391974, exporterPosition=1494387075}}
2022-07-12 01:26:19.339 CEST
RaftServer{raft-partition-partition-2}{role=FOLLOWER} - Delete existing log (lastIndex '494887084') and replace with received snapshot (index '494998838')
2022-07-12 01:26:19.773 CEST
Transition to FOLLOWER on term 19 requested.
...
2022-07-12 01:26:19.922 CEST
Transition to FOLLOWER on term 19 completed
```
2. Broker later becomes the leader
```
2022-07-12 01:26:37.184 CEST
Transition to LEADER on term 20 starting

2022-07-12 01:26:37.217 CEST
Recovered exporter 'Broker-1-Exporter-2' from snapshot at lastExportedPosition 1494387075
```
3. No new records are exported
```
2022-07-12 01:27:38.251 CEST
Unexpected exception occurred on periodically flushing bulk, will retry later.
2022-07-12 01:28:08.589 CEST
Error on exporting record with key 4503600188691487
2022-07-12 01:28:37.329 CEST
```
4. Tries to take snapshot at the same exportedPosition but with different processedPosition, but fails.
```
2022-07-12 01:28:37.329 CEST
Failed to take a snapshot for Broker-1-StreamProcessor-2
"io.camunda.zeebe.broker.system.partitions.NoEntryAtSnapshotPosition: Failed to take snapshot. Expected to find an indexed entry for determined snapshot position 1494387075 (processedPosition = 1494736551, exportedPosition=1494387075), but found no matching indexed entry which contains this position.
```

**Hypothesis**

It fails to get the index because, it is trying to get the "previous" entry -> entry at index `494998838`. Since there is already a snapshot at that index, raft leader would have only replicated event from index `494998839`. The record with the position `1494387075` is at index `494998839`.  So there is no inconsistency in the logs but it cannot take snapshot because the previous index is already compacted. 

**Expected behavior**

Leader should take a snapshot if exporter position has not changed, but it has processed new events. 

**Log/Stacktrace**

[Logs](https://console.cloud.google.com/logs/query;cursorTimestamp=2022-07-11T23:26:37.198702193Z;pinnedLogId=2022-07-11T23:28:37.329086224Z%2Fvhzpf3bafzqlmn73;query=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.pod_name%3D%22medic-cw-27-56ad2b36c8-benchmark-zeebe-1%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.namespace_name%3D%22medic-cw-27-56ad2b36c8-benchmark%22%0Aresource.labels.pod_name%3D%22medic-cw-27-56ad2b36c8-benchmark-zeebe-1%22%0AjsonPayload.context.partitionId%3D%222%22%0Atimestamp%3D%222022-07-11T23:28:37.329086224Z%22%0AinsertId%3D%22vhzpf3bafzqlmn73%22;summaryFields=jsonPayload%252Fcontext%252FpartitionId:false:32:beginning:false;timeRange=2022-07-11T22:59:07.329Z%2F2022-07-11T23:59:07.329Z?project=zeebe-io)

<details><summary>Full Stacktrace</summary>
 <p>

```
"io.camunda.zeebe.broker.system.partitions.NoEntryAtSnapshotPosition: Failed to take snapshot. Expected to find an indexed entry for determined snapshot position 1494387075 (processedPosition = 1494736551, exportedPosition=1494387075), but found no matching indexed entry which contains this position.
	at io.camunda.zeebe.broker.system.partitions.impl.StateControllerImpl.takeTransientSnapshotInternal(StateControllerImpl.java:160) ~[zeebe-broker-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.broker.system.partitions.impl.StateControllerImpl.lambda$takeTransientSnapshot$0(StateControllerImpl.java:69) ~[zeebe-broker-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:79) ~[zeebe-scheduler-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:44) ~[zeebe-scheduler-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:116) ~[zeebe-scheduler-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:97) ~[zeebe-scheduler-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:81) ~[zeebe-scheduler-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:189) ~[zeebe-scheduler-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
```

</p>
</details>

### PR DESCRIPTION
**Pull request by lenaschoenburg**
When figuring out where to take the next snapshot, we determine the snapshot position as the minimum of processing and exporter positions.

There was an edge case where a leader could process but no export.
In that case it'd use the exporter position as snapshot position and tryand find a log entry at that position.
If the log starts with the exporter position, for example because the same broker previously received a snapshot and compacted the log, no entry could be found which led to a failed snapshot.

We now try and use the latest snapshot's term and index if the log entry could not be found. This ensures that new snapshots can be taken even if nothing was exported since the last snapshot.

closes #9761

### CODE REVIEW

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/StateControllerImpl.java [hunk 1]
```diff
@@ -247,14 +263,6 @@ private void takeSnapshot(
...
-      final long lowerBoundSnapshotPosition, final long exportedPosition) {
-    final long snapshotPosition = Math.min(exportedPosition, lowerBoundSnapshotPosition);
-    LOG.trace(
-        "Based on lowest exporter position '{}' and last processed position '{}', determined '{}' as snapshot position.",
-        exportedPosition,
-        lowerBoundSnapshotPosition,
-        snapshotPosition);
-    return snapshotPosition;
-  }
+  private record NextSnapshotId(
```
Comment (deepthidevaki) [inline broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/StateControllerImpl.java:266]: :+1:

**Review by deepthidevaki — CHANGES_REQUESTED** [review]

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/StateControllerImpl.java [hunk 2]
```diff
@@ -192,6 +164,50 @@ private void takeTransientSnapshotInternal(
...
+    }
+
+    // No log entry for snapshot position - try to use the index and term of the last snapshot to
+    // take new one
+    final var latestSnapshot = constructableSnapshotStore.getLatestSnapshot();
+    if (latestSnapshot.isPresent()) {
+      LOG.warn(
+          "No log entry for next snapshot position {}, using index and term from previous snapshot",
+          snapshotPosition);
+      return new NextSnapshotId(
```
[the remark this answers is not in the document]
Comment (oleschoenburg) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/StateControllerImpl.java:198]: As discussed, this is not strictly necessary and it's okay to not do this check here.

[the remark this answers is not in the document]
Comment (deepthidevaki) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/StateControllerImpl.java:198]: So in that case it fails with `SnapshotAlreadyExists` error, which is ok.

**Review by deepthidevaki — APPROVED** [review]
Thanks. :tada:

**Pull request merged**